File: DMOJ/ECOO/Ring.py
# ECOO '20 P4 - Ring


def parse():
    n = int(input())
    instructions = []
    func_ptr = {}
    prev_jmp = []
    popped = 0
    for i in range(n):
        line = input()
        ptr = i - popped
        if line[0] == "F":
            instructions.append(["JMP"])
            prev_jmp.append(ptr)
            func_ptr[line.split()[-1]] = ptr + 1
        elif line[0] == "E":
            instructions[prev_jmp[-1]].append(ptr + 1)
            line += " 0"
            prev_jmp.pop()
            instructions.append(line.split())
        elif line[0] == "C":
            a, b = line.split()
            instructions.append([a, func_ptr[b]])
        else:
            instructions.append(line.split())

        # compress instructions
        if len(instructions) >= 2:
            if instructions[-1][0] == "ADD" and instructions[-2][0] == "ADD" or instructions[-1][0] != "ADD" and instructions[-1][0] == "SUB" and instructions[-2][0] != "ADD" and instructions[-2][0] == "SUB":
                instructions[-2][1] = int(instructions[-2][1]) + int(instructions[-1][1])
                instructions[-2][1] %= int(1e9) + 7
                instructions.pop()
                popped += 1
            elif instructions[-1][0] == "ADD" and instructions[-2][0] == "SUB":
                instructions[-2][0] = "ADD"
                instructions[-2][1] = -int(instructions[-2][1]) + int(instructions[-1][1])
                instructions[-2][1] %= int(1e9) + 7
                instructions.pop()
                popped += 1
            elif instructions[-1][0] != "ADD" and (instructions[-1][0] != "SUB" or instructions[-2][0] == "ADD") and instructions[-1][0] == "SUB":
                instructions[-2][1] = int(instructions[-2][1]) - int(instructions[-1][1])
                instructions[-2][1] %= int(1e9) + 7
                instructions.pop()
                popped += 1
            elif instructions[-1][0] == instructions[-2][0] == "MULT":
                instructions[-2][1] = int(instructions[-2][1]) * int(instructions[-1][1])
                instructions[-2][1] %= int(1e9) + 7
                instructions.pop()
                popped += 1
    return instructions


def execute(instructions):
    x = 0
    call_stack = [0]
    while call_stack[0] < len(instructions):
        ins, param = instructions[call_stack[-1]]
        if ins[0] == "J":
            call_stack[-1] = int(param)
        elif ins[0] == "A":
            x += int(param)
            call_stack[-1] += 1
            x %= int(1e9) + 7
        elif ins[0] == "S":
            x -= int(param)
            call_stack[-1] += 1
        elif ins[0] == "M":
            x *= int(param)
            call_stack[-1] += 1
            x %= int(1e9) + 7
        elif ins[0] == "C":
            call_stack.append(param)
        elif ins[0] == "E":
            call_stack.pop()
            call_stack[-1] += 1
    return x


for t in range(int(input())):
    print(execute(parse()))

File: DMOJ/ECOO/Ring.cpp
// ECOO '20 P4 - Ring

#include <bits/stdc++.h>
using namespace std;
typedef long long ll;

void mod(ll &a, ll b) {
    a = (b + (a % b)) % b;
}
ll solve() {
    // parse
    const ll MOD = 1e9 + 7;
    int n;
    cin >> n;
    vector<pair<char, ll>> instructions;
    unordered_map<string, int> func_ptr;
    stack<int> prev_jmp;
    string cmd, param;
    int popped = 0;
    for (int i = 0; i < n; i++) {
        cin >> cmd;
        char c = cmd[0];

        if (c != 'E') cin >> param;

        int ptr = i - popped;
        switch (c) {
            case 'F':
                instructions.emplace_back('J', 0);
                prev_jmp.push(ptr);
                func_ptr[param] = ptr + 1;
                break;
            case 'E':
                instructions[prev_jmp.top()].second = ptr + 1;
                prev_jmp.pop();
                instructions.emplace_back(c, 0);
                break;
            case 'C':
                instructions.emplace_back(c, func_ptr[param]);
                break;
            case 'S':
                instructions.emplace_back('A', -stoi(param));
                break;
            default:
                instructions.emplace_back(c, stoi(param));
        }

        if (instructions.size() < 2) continue;
        int l = instructions.size();
        if (instructions[l - 1].first == 'A' && instructions[l - 2].first == 'A') {
            instructions[l - 2].second += instructions[l - 1].second;
            mod(instructions[l - 2].second, MOD);
            instructions.pop_back();
            popped++;
        } else if (instructions[l - 1].first == 'M' && instructions[l - 2].first == 'M') {
            instructions[l - 2].second *= instructions[l - 1].second;
            mod(instructions[l - 2].second, MOD);
            instructions.pop_back();
            popped++;
        }
    }

    // execute
    ll x = 0;
    list<int> call_stack;
    call_stack.push_front(0);
    while (call_stack.front() < int(instructions.size())) {
        auto [c, v] = instructions[call_stack.back()];
        switch (c) {
            case 'J':
                call_stack.back() = v;
                break;
            case 'A':
                x += v;
                call_stack.back()++;
                mod(x, MOD);
                break;
            case 'M':
                x *= v;
                call_stack.back()++;
                mod(x, MOD);
                break;
            case 'C':
                call_stack.push_back(v);
                break;
            case 'E':
                call_stack.pop_back();
                call_stack.back()++;
                break;
        }
    }
    return x;
}
int main() {
    ios_base::sync_with_stdio(0);
    cin.tie(0);
    int t;
    cin >> t;
    while (t--) {
        cout << solve() << '\n';
    }
}
